File: src/main/java/com/smarthome/backend/server/api/modules/ModuleEventStreamManager.java
package com.smarthome.backend.server.api.modules;

/**
 * Interface für EventStreamManager von Modulen.
 * Module können dieses Interface implementieren, um Eventstreams für ihre Geräte zu verwalten.
 * 
 * Der zentrale EventStreamManager verwaltet mehrere ModuleEventStreamManager-Instanzen
 * und hält sie am Laufen.
 */
public interface ModuleEventStreamManager {
    
    /**
     * Startet den EventStream für dieses Modul.
     * Diese Methode sollte alle notwendigen Eventstreams für die Geräte des Moduls starten.
     * 
     * @throws Exception wenn der EventStream nicht gestartet werden kann
     */
    void start() throws Exception;
    
    /**
     * Stoppt den EventStream für dieses Modul.
     * Diese Methode sollte alle laufenden Eventstreams für die Geräte des Moduls stoppen.
     * 
     * @throws Exception wenn der EventStream nicht gestoppt werden kann
     */
    void stop() throws Exception;
    
    /**
     * Gibt zurück, ob der EventStream aktuell läuft.
     * 
     * @return true wenn der EventStream läuft, false sonst
     */
    boolean isRunning();
    
    /**
     * Gibt die Modul-ID zurück, zu der dieser EventStreamManager gehört.
     * 
     * @return Die Modul-ID (z.B. "hue", "denon", "matter")
     */
    String getModuleId();
    
    /**
     * Gibt die Manager-ID zurück, die diesen EventStreamManager eindeutig identifiziert.
     * Diese ID kann verwendet werden, um mehrere EventStreamManager-Instanzen
     * für dasselbe Modul zu unterscheiden.
     * 
     * @return Die Manager-ID (z.B. "hue-bridge-1", "denon-speaker-2")
     */
    String getManagerId();
    
    /**
     * Gibt eine Beschreibung des EventStreamManagers zurück (für Logging/Debugging).
     * 
     * @return Eine Beschreibung des EventStreamManagers
     */
    String getDescription();
}

File: src/main/node/com/smarthome/backend/server/api/modules/moduleEventStreamManager.ts
import { logger } from "../../../logger.js";
import { runWithSource, EventSource } from "../../events/EventSource.js";
import { ModuleEvent } from "./moduleEvent.js";
import { ModuleEventController } from "./moduleEventController.js";
import { DeviceManager } from "../entities/devices/deviceManager.js";

export abstract class ModuleEventStreamManager<C extends ModuleEventController, E extends ModuleEvent> {
  protected managerId: string;
  protected moduleId: string;
  protected controller: C;
  protected deviceManager: DeviceManager;
  private running = false;

  constructor(managerId: string, moduleId:string, controller: C, deviceManager: DeviceManager) {
    this.managerId = managerId;
    this.moduleId = moduleId;
    this.controller = controller;
    this.deviceManager = deviceManager;
  }

  protected abstract handleEvent(event: E): void;
  protected abstract startEventStream(callback: (event: E) => void): Promise<void>;
  protected abstract stopEventStream(): Promise<void>;

  public async start(): Promise<void> {
    if (this.running) {
      logger.warn(this.managerId + " EventStream laeuft bereits");
      return;
    }
    this.running = true;
    const callback: (event: E) => void = (event: E) => runWithSource(EventSource.SENSOR, () => this.handleEvent(event));
    // Start asynchronously, damit die Event-Loop nicht blockiert und z. B. HTTP-Anfragen weiter bedient werden
    void this.startEventStream(callback)
      .then(() => logger.info(this.managerId + " EventStream gestartet"))
      .catch(err => {
        this.running = false;
        logger.error({ err, managerId: this.managerId }, "Fehler beim Starten des EventStreams");
      });
  }

  public async stop():Promise<void> {
    if (!this.running) return;
    this.running = false;
    await this.stopEventStream();
    logger.info(this.managerId+" EventStream gestoppt");
  }

  public getManagerId(): string {
    return this.managerId;
  }

  public getModuleId(): string {
    return this.moduleId;
  }

  public getDescription():string {
    return `${this.moduleId} EventStream fuer Manager ${this.managerId}`;
  }

  public isRunning(): boolean {
    return this.running;
  }
}

